plugins/inventory/falcon_discover.py

```diff
@@ -0,0 +1,378 @@
+        cache_needs_update = user_cache_setting and not cache
+
+        # attempt to read the cache if inventory isn't being refreshed and the user has caching enabled
+        if attempt_to_read_cache:
+            try:
+                host_details = self._cache[cache_key]
+            except KeyError:
+                # This occurs if the cache_key is not in the cache or if the cache_key expired, so the cache needs to be updated
+                cache_needs_update = True
+        if not attempt_to_read_cache or cache_needs_update:
+            # parse the provided inventory source
+            falcon = self._authenticate()
+            # Get the filter expression
+            fql = self.get_option("filter")
+            # Get host details
+            host_details = self._get_host_details(falcon, fql)
+        if cache_needs_update:
+            self._cache[cache_key] = host_details
+
+        # Add hosts to inventory
+        self._add_host_to_inventory(host_details)
+
+    def _credential_setup(self):
+        """Setup credentials for FalconPy."""
+        cred_mapping = {
+            "client_id": "FALCON_CLIENT_ID",
+            "client_secret": "FALCON_CLIENT_SECRET",
+            "member_cid": "FALCON_MEMBER_CID",
+            "cloud": "FALCON_CLOUD",
+        }
+
+        creds = {}
+        for key, env in cred_mapping.items():
+            value = self.get_option(key) or os.getenv(env)
+            if value:
+                if key == "cloud":
+                    creds["base_url"] = value
+                else:
+                    creds[key] = value
+
+        # Make sure we have client_id and client_secret
+        if "client_id" not in creds or "client_secret" not in creds:
+            raise ValueError(
+                "You must provide a client_id and client_secret to authenticate to the Falcon API."
+            )
+
+        return creds
+
+    def _authenticate(self):
+        """Authenticate to the CrowdStrike Falcon API."""
+        creds = self._credential_setup()
+
+        return Discover(**creds)
+
+    def _get_host_details(self, falcon, fql):
+        """Query hosts from Falcon Discover."""
+        max_limit = 100  # Maximum limit allowed by Falcon Discover
+        host_ids = []
+        host_details = []
+        running = True
+        offset = None
+        while running:
+            host_lookup = falcon.query_hosts(filter=fql, offset=offset, limit=max_limit)
+            if host_lookup["status_code"] != 200:
+                raise SystemExit(
+                    f"Unable to query hosts: {host_lookup['body']['errors']}"
+                )
+
+            if host_lookup["body"]["resources"]:
+                host_ids = host_lookup["body"]["resources"]
+            else:
+                raise SystemExit(
+                    f"No hosts were identified by the filter expression: {fql}"
```

Comment:
Not sure if this should be fatal. Imagine nightly job to remediate unmanaged hosts. If there are no hosts, that's fine.

Reply:
Good catch - that makes sense. We should just return back host_details at this point